tethysapp/temp_precip_trends/functions.py

```diff
@@ -0,0 +1,192 @@
+import json
+import numpy as np
+import xarray as xr
+import pandas as pd
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+
+from .app import TempPrecipTrends as app
+
+
+def get_data(variable, params):
+    """
+    Calculate cumulative precipitation over given time.
+
+    Args:
+        variable(str): Name of the variable to query.
+        params(Python Dict): Python dictionary with request parameters.
+
+    Returns:
+        Python Dict: JSON-compatible Python Dict with specified variable data.
+    """
+    catalog = app.get_spatial_dataset_service(app.THREDDS_SERVICE_NAME, as_engine=True)
+    geometry = params['geometry']
+    start_time = params.get('start_time', None)
+    end_time = params.get('end_time', None)
+    vertical_level = params.get('vertical_level', None)
+
+    time_series = extract_time_series_at_location(
+        catalog=catalog,
+        geometry=geometry,
+        variable=variable,
+        start_time=start_time,
+        end_time=end_time,
+        vertical_level=vertical_level
+    )
+
+    return jsonify(time_series, variable)
+
+
+def get_cum_precip_data(params):
+    """
+    Calculate cumulative precipitation over given time.
+
+    Args:
+        params(Python Dict): Python dictionary with request parameters.
+
+    Returns:
+        Python Dict: JSON-compatible Python Dict with cummulative precipitation.
+    """
+    catalog = app.get_spatial_dataset_service(app.THREDDS_SERVICE_NAME, as_engine=True)
+    geometry = params['geometry']
+    start_time = params.get('start_time', None)
+    end_time = params.get('end_time', None)
+    vertical_level = params.get('vertical_level', None)
+
+    time_series = extract_time_series_at_location(
+        catalog=catalog,
+        geometry=geometry,
+        variable='sum_tp_mm',
+        start_time=start_time,
+        end_time=end_time,
+        vertical_level=vertical_level
+    )
+
+    cum_precip = time_series.sum_tp_mm.cumsum(dim='obs', skipna=True)
+    time_series['cum_pr_mm'] = cum_precip
+
+    return jsonify(time_series, 'cum_pr_mm')
+
+
+def jsonify(dataset, variable):
+    """
+    Extract a time series from a THREDDS dataset at the given location.
+
+    Args:
+        dataset(xarray.Dataset): an xarray dataset.
+        variable(str): Name of the variable to query.
+
+    Returns:
+        Python Dict: JSON-compatible Python Dict.
+    """
+    df = pd.DataFrame(data={variable: np.transpose(dataset[variable].data)}, index=dataset.time.data)
+    df.index = df.index.strftime('%Y-%m-%dT%H:%M:%SZ')
+    df.index.name = 'datetime'
+
+    context = {
+        'time_series': {
+            'datetime': df.index.tolist(),
+            variable: df[variable].to_list()
+        }
+    }
+
+    context['time_series'].update(df.to_dict(orient='list'))
```

Comment:
this is kind of a safety net. dict.update() only updates stuff if for whatever reason the keys are not there or the values are different. I removed it because our returned df only has one variable so it's was never going to do anything.